File: common/src/main/kotlin/net/impleri/playerskills/skills/registry/Players.kt
package net.impleri.playerskills.skills.registry

import net.impleri.playerskills.api.Skill
import net.impleri.playerskills.api.SkillType
import net.impleri.playerskills.skills.registry.storage.SkillStorage.read
import net.impleri.playerskills.skills.registry.storage.SkillStorage.write
import net.impleri.playerskills.utils.PlayerSkillsLogger
import net.minecraft.resources.ResourceLocation
import java.util.UUID

object Players {
  /**
   * MUTABLE in-memory cache
   */
  private val players: MutableMap<UUID, List<Skill<*>>> = HashMap()

  /**
   * Helper function to save to in-memory cache only
   */
  private fun saveToCache(playerId: UUID, skills: List<Skill<*>>) {
    players[playerId] = skills
  }

  /**
   * Helper function to save to in-memory cache AND persistent storage
   */
  private fun save(playerId: UUID, skills: List<Skill<*>>) {
    saveToCache(playerId, skills)
    writeToStorage(playerId, skills)
  }

  /**
   * Get a copy of the in memory cache
   */
  fun entries(): Map<UUID, List<Skill<*>>> {
    return players.toMap()
  }

  fun has(playerId: UUID): Boolean {
    return players.containsKey(playerId)
  }

  /**
   * Get all skills for a player from the in-memory cache
   */
  fun get(playerId: UUID): List<Skill<*>> {
    return players[playerId] ?: return readFromStorage(playerId)
  }

  private fun handleOpenFor(playerId: UUID): List<Skill<*>> {
    PlayerSkillsLogger.SKILLS.info("Opening player $playerId, ensuring saved skills are still valid")

    // Get all the names of the registered skills
    val registeredSkillNames = Skills.entries()
      .map { it.name }

    val rawSkills = this.get(playerId)

    // Get an intersection of saved skills that are still registered
    val savedSkills = rawSkills
      .filter { registeredSkillNames.contains(it.name) }

    val savedSkillNames = savedSkills
      .map { it.name }

    val newSkills = Skills.entries()
      .filter { !savedSkillNames.contains(it.name) }

    return savedSkills + newSkills
  }

  /**
   * Instantiates the in-memory cache for a player. Note that this will automatically prune saved skills that do not
   * match the skill type in the Skills Registry as well as those which no longer exist at all in the Skills Registry
   */
  fun open(playerId: UUID) {
    val skills = handleOpenFor(playerId)

    // Immediately sync in-memory cache AND persistent storage with updated skills set
    save(playerId, skills)
  }

  fun open(playerIds: List<UUID>): List<UUID> {
    val playersAdded = playerIds.filter { !players.containsKey(it) }

    val skillsList = playerIds.associateWith { handleOpenFor(it) }

    players.putAll(skillsList)

    return playersAdded
  }

  /**
   * Helper function to upsert a skill for a player in memory only
   */
  private fun handleUpsert(playerId: UUID, skill: Skill<*>): List<Skill<*>> {
    // Only replace skill with same name AND type (edge case of same name but different type is handled in openPlayer)
    val filteredSkills = this.get(playerId).filter { skill.name != it.name }

    val addedSkills = listOf(skill)

    val newSkills = filteredSkills + addedSkills

    saveToCache(playerId, newSkills)

    return newSkills
  }

  /**
   * Upsert a skill for a player and saves to persistent storage
   */
  fun upsert(playerId: UUID, skill: Skill<*>): List<Skill<*>> {
    val newSkills = handleUpsert(playerId, skill)

    save(playerId, newSkills)

    return newSkills
  }

  /**
   * Add a skill to a player only if the player does not have it
   */
  fun add(playerId: UUID, skill: Skill<*>): List<Skill<*>> {
    val existingSkills = this.get(playerId)

    // Skill already exists, so do nothing
    if (existingSkills.any { skill == it }) {
      return existingSkills
    }
    val newSkills = handleUpsert(playerId, skill)
    save(playerId, newSkills)
    return newSkills
  }

  /**
   * Completely remove a skill from a player
   */
  fun remove(playerId: UUID, name: ResourceLocation): List<Skill<*>> {
    val newSkills = this.get(playerId).filter { it.name !== name }

    save(playerId, newSkills)

    return newSkills
  }

  private fun handleCloseFor(playerId: UUID) {
    PlayerSkillsLogger.SKILLS.info("Closing player $playerId, ensuring skills are saved")

    val skills = this.get(playerId)

    writeToStorage(playerId, skills)
  }

  /**
   * Save a player's skills to persistent storage then remove player from in-memory cache
   */
  fun close(playerId: UUID) {
    handleCloseFor(playerId)

    players.remove(playerId)
  }

  fun close(playerIds: List<UUID>) {
    playerIds.forEach { handleCloseFor(it) }
    playerIds.forEach { players.remove(it) }
  }

  /**
   * Save multiple players' skills to persistent storage then remove them from in-memory cache
   */
  fun close(): List<UUID> {
    val playerIds = players.keys.toList()
    close(playerIds)

    // Wipe away any players that are somehow still present
    players.clear()
    return playerIds
  }

  private fun readFromStorage(playerId: UUID): List<Skill<*>> {
    PlayerSkillsLogger.SKILLS.debug("Restoring saved skills for $playerId")
    return read(playerId).mapNotNull { SkillType.unserializeFromString(it) }
  }

  private fun writeToStorage(playerId: UUID, skills: List<Skill<*>>) {
    PlayerSkillsLogger.SKILLS.debug("Saving skills for $playerId")

    val rawSkills: List<String> = skills
      .map { SkillType.serializeToString(it) }
      .filter { it.isNotEmpty() }

    write(playerId, rawSkills)
  }
}

File: common/src/main/scala/net/impleri/playerskills/skills/registry/Players.scala
package net.impleri.playerskills.skills.registry

import net.impleri.playerskills.api.skills.Skill
import net.impleri.playerskills.api.skills.SkillType
import net.impleri.playerskills.events.handlers.EventHandlers
import net.impleri.playerskills.skills.registry.storage.SkillFileMissing
import net.impleri.playerskills.skills.registry.storage.SkillStorage
import net.impleri.playerskills.utils.PlayerSkillsLogger
import net.minecraft.resources.ResourceLocation

import java.util.UUID
import scala.collection.mutable
import scala.util.chaining._

/**
 * Handles the cache side of interacting with player skills
 */
sealed trait CachedPlayers {
  private val players: mutable.Map[UUID, List[Skill[_]]] = mutable.HashMap()

  private[playerskills] def init(): Unit = ()

  def entries(): Map[UUID, List[Skill[_]]] = Map.from(players)

  def has(playerId: UUID): Boolean = players.contains(playerId)

  def readFromCache(playerId: UUID): Option[List[Skill[_]]] = players.get(playerId)

  protected def writeToCache(playerId: UUID, skills: List[Skill[_]]): Unit = players.update(playerId, skills)

  protected def writeBulkCache(values: Map[UUID, List[Skill[_]]]): List[UUID] =
    values
      .tap(players.addAll)
      .keys
      .toList

  protected def removeFromCache(playerId: UUID): Option[List[Skill[_]]] = players.remove(playerId)

  protected def removeBulkCache(playerIds: List[UUID]): Unit = playerIds.foreach(removeFromCache)

  protected def clearCache(): Unit = players.clear()
}

/**
 * Handles the I/O side of interacting with player skills
 */
trait PersistedPlayers {
  protected def storage: () => SkillStorage
  protected def readFromStorage(playerId: UUID): Option[List[Skill[_]]] =
    storage().read(playerId)
      .tap {
        case Right(_) => PlayerSkillsLogger.SKILLS.debug(s"Restoring saved skills for $playerId")
        case Left(SkillFileMissing(_)) => ()
        case Left(e) => PlayerSkillsLogger.SKILLS.warn(e.toString)
      }
      .map(SkillType.deserializeAll)
      .toOption

  private def serializeSkills(skills: List[Skill[_]]): List[String] = skills.flatMap(SkillType.serialize(_))

  protected def writeToStorage(playerId: UUID, skills: List[Skill[_]]): Boolean =
    storage().write(playerId, serializeSkills(skills))
      .tap {
        case Right(_) => PlayerSkillsLogger.SKILLS.info(s"Saving skills for $playerId")
        case Left(e) => PlayerSkillsLogger.SKILLS.warn(e.toString)
      }
      .toOption
      .getOrElse(false)
}

/**
 * Orchestrated handling of player skills
 */
object Players extends CachedPlayers with PersistedPlayers {
  override def storage: () => SkillStorage = () => EventHandlers.withServer.pipe(SkillStorage.apply)
  private def save(playerId: UUID, skills: List[Skill[_]]): Boolean = {
    writeToCache(playerId, skills)
    writeToStorage(playerId, skills)
  }

  private def filterRegisteredSkills(source: List[Skill[_]], target: List[Skill[_]]) =
    source
      .map(_.name)
      .pipe(s => target.filter(t => s.contains(t.name)))

  private def ensureRegisteredSkills(source: List[Skill[_]], target: List[Skill[_]]) =
    target
      .map(_.name)
      .pipe(t => source.filterNot(s => t.contains(s.name)))

  private def openFor(playerId: UUID) =
    readFromStorage(playerId)
      .getOrElse(List.empty)
      .pipe(filterRegisteredSkills(Skills.entries, _))
      .pipe(s =>  s ++ ensureRegisteredSkills(Skills.entries, s))
      .tap(save(playerId, _))

  def open(playerIds: List[UUID]): List[UUID] =
    playerIds
      .filterNot(has)
      .map(p => (p, openFor(p)))
      .toMap
      .pipe(writeBulkCache)

  def open(playerId: UUID): List[Skill[_]] =
    open(List(playerId))
      .headOption
      .flatMap(readFromCache)
      .getOrElse(List.empty)

  def get(playerId: UUID): List[Skill[_]] =
    readFromCache(playerId)
      .getOrElse(open(playerId))

  def upsert(playerId: UUID, skill: Skill[_]): List[Skill[_]] =
    get(playerId)
      .filterNot(_.name == skill.name)
      .tap(ss => if (ss.nonEmpty) PlayerSkillsLogger.SKILLS.info(s"Replacing ${skill.name} for $playerId"))
      .pipe(_ ++ List(skill))
      .tap(writeToStorage(playerId, _))

  private def safeAdd(skill: Skill[_])(skills: List[Skill[_]]): List[Skill[_]] =
    if (skills.exists(_.name == skill.name)) skills else skills ++ List(skill)

  def addSkill(playerId: UUID, skill: Skill[_]): List[Skill[_]] =
    get(playerId)
      .pipe(safeAdd(skill))
      .tap(save(playerId, _))

  def removeSkill(playerId: UUID, name: ResourceLocation): List[Skill[_]] =
    get(playerId)
      .pipe(_.filterNot(_.name == name))
      .tap(save(playerId, _))

  def removeSkill(playerId: UUID, skill: Skill[_]): List[Skill[_]] = removeSkill(playerId, skill.name)

  private def closeFor(playerId: UUID) =
    playerId
      .tap(p => PlayerSkillsLogger.SKILLS.info(s"Closing player $p, ensuring skills are saved"))
      .pipe(get)
      .pipe(writeToStorage(playerId, _))

  def close(playerIds: List[UUID]): List[UUID] =
    playerIds
      .map(p => (p, closeFor(p)))
      .toMap
      .partition(_._2 == false)
      .tap(p => if(p._1.nonEmpty) PlayerSkillsLogger.SKILLS.warn(s"Could not save player data for: ${p._1.keys.mkString(",")}"))
      ._2
      .keys
      .toList
      .tap(removeBulkCache)

    def close(playerId: UUID): Boolean =
      close(List(playerId))
        .contains(playerId)


  def close(): List[UUID] =
    entries()
      .pipe(_.keys.toList)
      .pipe(close)
      .tap(_ => clearCache())
}
